File: sliding_window_maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:

        res = []
        left = right = 0
        q = deque()

        while right < len(nums):
            while q and nums[right] > nums[q[-1]]:
                q.pop()
            q.append(right)

            if left > q[0]:
                q.popleft()
            
            if right + 1 >= k:
                res.append(nums[q[0]])
                left += 1
            right += 1
        
        return res

File: sliding_window_maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        List<Integer> res = new ArrayList<>();
        int left = 0;
        int right = 0;
        Deque<Integer> q = new ArrayDeque<>();

        while (right < nums.length) {
            while (!q.isEmpty() && nums[right] > nums[q.peekLast()]) {
                q.pollLast();
            }
            q.offerLast(right);

            if (left > q.peekFirst()) {
                q.pollFirst();
            }

            if (right + 1 >= k) {
                res.add(nums[q.peekFirst()]);
                left++;
            }
            right++;
        }

        return res.stream().mapToInt(Integer::intValue).toArray();        
    }
}
